frontend/src/metabase/common/components/EntityPicker/components/EntityPickerModal/EntityPickerModal.tsx

```diff
@@ -232,14 +234,26 @@ export function EntityPickerModal<
     [folderModels, selectedTabId, onItemSelect],
   );
 
+  const handleQueryChange = useCallback(
+    (newSearchQuery: string) => {
+      setSearchQuery(newSearchQuery);
+
+      // automatically switch to search tab when it becomes available
+      if (!searchQuery && newSearchQuery) {
+        setSelectedTabId(SEARCH_TAB_ID);
+      }
+
+      // restore previous tab when clearing search while on search tab
+      if (searchQuery && !newSearchQuery && selectedTabId === SEARCH_TAB_ID) {
+        setSelectedTabId(previousTabId ?? initialTabId);
+      }
+    },
+    [searchQuery, selectedTabId, previousTabId, initialTabId],
+  );
+
   useEffect(() => {
-    // when the searchQuery changes, switch to the search tab
-    if (searchQuery) {
-      setSelectedTabId(SEARCH_TAB_ID);
-    } else {
-      setSelectedTabId(initialTabId);
-    }
-  }, [searchQuery, initialTabId]);
+    setSelectedTabId(initialTabId);
```

Comment:
I changed the effect into an imperative code because the hook started to have so many deps, that it was difficult to make it run only when we want it to.

However, I couldn't fully remove the effect.
I'm not sure if the effect was originally intended to do this, but it already was doing it, and it needs to stay like this (otherwise we open wrong tab initially in some cases).